build/pluginctl/logs.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+const (
+	logsPerPage     = 100 // logsPerPage is the number of log entries to fetch per API call
+	timeStampFormat = "2006-01-02 15:04:05.000 Z07:00"
+)
+
+// logs fetches the latest 500 log entries from Mattermost,
+// and prints only the ones related to the plugin to stdout.
+func logs(ctx context.Context, client *model.Client4, pluginID string) error {
+	for i := 5 - 1; i >= 0; i-- {
+		logs, err := fetchLogs(ctx, client, i, pluginID, time.Unix(0, 0))
+		if err != nil {
+			return fmt.Errorf("failed to fetch log entries: %w", err)
+		}
+
+		err = printLogEntries(logs)
+		if err != nil {
+			return fmt.Errorf("failed to print logs entries: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// watchLogs fetches log entries from Mattermost and print them to stdout.
+// It will return without an error when ctx is canceled.
+func watchLogs(ctx context.Context, client *model.Client4, pluginID string) error {
+	now := time.Now()
+	var oldest string
+
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			for {
+				logs, err := fetchLogs(ctx, client, 0, pluginID, now)
+				if err != nil {
+					return fmt.Errorf("failed to fetch log entries: %w", err)
+				}
+
+				var allNew bool
+				logs, oldest, allNew = checkOldest(logs, oldest)
+
+				err = printLogEntries(logs)
+				if err != nil {
+					return fmt.Errorf("failed to print logs entries: %w", err)
+				}
+
+				if !allNew {
+					// No more logs to fetch
+					break
+				}
+			}
+		}
+	}
+}
+
+// checkOldest check a if logs contains new log entries.
+// It returns the filtered slice of log entries, the new oldest entry and whether or not a entries were new.
+func checkOldest(logs []string, oldest string) ([]string, string, bool) {
+	if len(logs) == 0 {
+		return nil, oldest, false
+	}
+
+	i := slices.Index(logs, oldest)
+	switch i {
+	case -1:
+		// Every log entry is new
+		return logs, logs[(len(logs) - 1)], true
+	case len(logs) - 1:
+		// No new log entries
+		return nil, oldest, false
+	default:
+		// Filter out oldest log entry
+		return logs[i+1:], logs[(len(logs) - 1)], false
+	}
+}
+
+// fetchLogs fetches log entries from Mattermost
+// and filters them based on pluginID and timestamp.
+func fetchLogs(ctx context.Context, client *model.Client4, page int, pluginID string, since time.Time) ([]string, error) {
+	logs, _, err := client.GetLogs(ctx, page, logsPerPage)
```

Comment:
Not required for this PR, but I wonder if we can do better than just polling a logs endpoint that may not give us all the logs if a high volume of logs gets emitted? (e.g. 101 new log lines between polling intervals means some logs get "lost").

When I develop and want plugin logs, I basically just `tail -f | jq ...` the log file. Could we do something as reliable as that?

Reply:
Regarding the high volume of logs, please see https://github.com/mattermost/mattermost-plugin-starter-template/pull/193#discussion_r1441734980. The code does now fetch more then one page.

Reply:
> When I develop and want plugin logs, I basically just `tail -f | jq ...` the log file. Could we do something as reliable as that?

I would love to use this simple, reliable solution, but I wanted to build a tool that also works for remote servers like the ones in MM Cloud. Is there another similar tool that we can use here instead of reading the local file?

Reply:
For MM Cloud, at least, I just use Grafana's livestreaming feature. Unless we build something more robust into the server, I guess we'll be limited in what we can do here, which might be fine for the level of logging needed.